### PR TITLE
IDE-2850: React as external library + fix several issues in n4jsx transpiler

### PR DESCRIPTION
**Pull request by qtran-n4**

### CODE REVIEW

**Comment by qtran-n4** [comment]
@mor-n4 : Please review the use cases you described the user story.
@yoosiba: Please review the fixes in the JSXBackend.

Thanks!

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/helpers/ReactHelper.xtend [hunk 1]
```diff
@@ -73,27 +78,32 @@ class ReactHelper {
 	def private TClassifier lookUpReactClassifier(EObject context, String reactClassifierName) {
 		val String key = REACT_KEY + "." + reactClassifierName;
 		return resourceScopeCacheHelper.get(key, context.eResource, [
-			val project = n4jsCore.findProject(context.eResource.URI).get;
+			val index = resourceDescriptionsProvider.getResourceDescriptions(context.eResource)
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/helpers/ReactHelper.xtend]: use IN4JSCore to get the index

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		// nop
+	}
+
+	/** Make sure the alias to react module is React. */
+	@Check
+	def checkReactImport(NamespaceImportSpecifier importSpecifier) {
+		val module = importSpecifier.importedModule
+		if (module !== null && REACT_PROJECT_ID.equalsIgnoreCase(module.projectId)) {
+			if (importSpecifier.alias != REACT_ALIAS) {
+						addIssue(
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend]: IMHO forcing `React` is misinterpretation of JSX.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend [hunk 3]
```diff
@@ -256,24 +256,25 @@ class ModuleWrappingTransformation extends Transformation {
 
 				val module = state.info.getImportedModule(elementIM);
 
+				val isReactImport = JSXBackendHelper.isJsxBackendImportDeclaration(elementIM)
+
 				// calculate names in output
 				val completeModuleSpecifier =
-					if (JSXBackendHelper.isJsxBackendModule(module)) {
+					if (isReactImport) {
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: There are other implementations of JSX, not just `React`.
Why change the name? This name suggests concrete implementation, instead intended semantics.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend [hunk 4]
```diff
@@ -283,6 +284,10 @@ class ModuleWrappingTransformation extends Transformation {
 					completeModuleSpecifier
 				};
 
+				if (isReactImport && !actualModuleSpecifier.startsWith('@@cjs')) {
```
Comment (yoosiba) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: Don't we have constant for this somewhere?

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/messages.properties [hunk 5]
```diff
@@ -26,3 +26,4 @@ JSXSPREADATTRIBUTE_WRONG_SUBTYPE=error;;;Attribute {0} has wrong type because {1
 JSXSPREADATTRIBUTE_NOT_DECLARED_IN_PROPS=warning;;;Attribute {0} is not a declared property in the props of {1}.
 # 0: The name of attribute in JSX property, 1: name of JSX element
 JSXSPROPERTYATTRIBUTE_NOT_DECLARED_IN_PROPS=warning;;;Attribute {0} is not a declared property in the props of {1}.
+REACT_ALIAS_NOT_ALLOWED=error;;;Alias to react must be React.
```
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/messages.properties]: it's a namespace, at least in cases as:
```
import * as React from ...
```

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend (hunk 4 above)
[the remark this answers is not in the document]
Comment (mor-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: alternative: add import statement in `SanitizeImportsTransformation`;
then existing `ModuleWrappingTransformation` will handle `@@cjs` as usual.

**Comment by mor-n4** [comment]
also: compare react look-up in `ReactHelper` with scoping of named imports

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/helpers/ReactHelper.xtend (hunk 1 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/helpers/ReactHelper.xtend]: done

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend (hunk 2 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend]: You are right that there could be multiple JSX backends and we shouldn't hard-code React. 
I discussed this with @dbo  and he said we could leave it like this for now and in the future if we need to support multiple JSX backends, then we can improve this.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend (hunk 3 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: See comment above

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend (hunk 4 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: I have refactored the code. This string goes away.

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/messages.properties (hunk 5 above)
[the remark this answers is not in the document]
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/messages.properties]: Done

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend (hunk 4 above)
[the remark this answers is not in the document]
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/ModuleWrappingTransformation.xtend]: Done

File: plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/n4jsx/transpiler/utils/JSXBackendHelper.java [hunk 6]
```diff
@@ -84,22 +87,19 @@ public String getBackendElementFactoryMethodName() {
...
-			return false;
-		}
-		return module.getQualifiedName().endsWith(JSX_BACKEND_MODULE_NAME);
-	}
-
 	/** Checks if given import declaration looks like JSX backend import, e.g. "(...) from "react" */
 	public static boolean isJsxBackendImportDeclaration(ImportDeclaration declaration) {
-		return isJsxBackendModule(declaration.getModule());
+		return declaration.getImportSpecifiers().stream().anyMatch(specifier -> isJsxBackendImportSpecifier(specifier));
+		// return isJsxBackendModule(declaration.getModule());
```
[the remark this answers is not in the document]
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/n4jsx/transpiler/utils/JSXBackendHelper.java]: done

File: plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	 */
+	override register(EValidatorRegistrar registrar) {
+		// nop
+	}
+
+	/** Make sure the alias to react module is React. */
+	@Check
+	def checkReactImport(NamespaceImportSpecifier importSpecifier) {
+		val module = importSpecifier.importedModule
+		if (module !== null && REACT_PROJECT_ID.equalsIgnoreCase(module.projectId)) {
```
[the remark this answers is not in the document]
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.n4jsx/src/org/eclipse/n4js/n4jsx/validation/N4JSXImportValidator.xtend]: Done. Move this check into ReactHelper

**Comment by qtran-n4** [comment]
@yoosiba @mor-n4: I have incorporated your comments. 
@mor-n4 : ReactHelper is used to look up stuff that is used during scoping. As a result, we cannot use scoping for React lookup. Otherwise I would have cash 22...

**Comment by qtran-n4** [comment]
@mor-n4 If you are satisfied with the changes, please approve the PR.

**Comment by qtran-n4** [comment]
Builds are green.

**Comment by mor-n4** [comment]
this is blocked by #355

**Comment by mor-n4** [comment]
created follow-up issue #358

**Comment by qtran-n4** [comment]
Thanks for your reviews!

**Pull request merged**